Revamp AES Part 1; including random & encrypted IV, hashed key and max keysize

Logan007 opened the pull request:
This pull request addresses some security issues in the AES module _transform_aes_ as follows:

**Key Stretching**
- the key provided by the user gets hashed
- this hash is used to derive the actual encryption and decryption key
- also, a secondary key is derived from the hash
- for being able to derive two independent keys in the same step, a hash function with wider output is used (SHA512)

**CBC data encryption**
- make use of an full block sized initialization vector (IV) instead of a nonce
- the IV is randomly set for every packet
- as the first data block is quite predictable (preamble, NIC's MACs), the IV gets encrypted as single block using the secondary key
- replay protection could be added to the IV and may be considered in Part 2

**Message authentication**
- a field for a message authentication code is filled with a ECBC-MAC, i.e. the last cipher block gets encrypted using the secondary key
- to ensure safety of ECBC-MAC, i.e. keep it prefix-free, data length is prepended to the data to be encrypted
- thus, the number of padded bytes at the last packet's end is not neccessary anymore and gets removed
- verification of the message authentication code was not first priority and may be included in Part 2

**Any other business**
- version number increased to 2 as packet design changed
- fixes a packet loss issue - at least for local edges
- author's first pull request ever - please be kind :)

Commit:
Revamp AES Part 1; including random & encrypted IV, hashed key and max keysize

emanuele-f commented:
Thank you! The pull request is very clean, your contribution is very welcome. I'm adding some inline comments for the review. If possible, also provide some links to online resources which explain the concepts used, for example the "prefix free" part of ECBC-MAC. Here is a reminder for me of the encryption/decryption process (please correct me if something is wrong).

```
*Encryption*:
iv = randomIv()
aes_cbc_encrypt(iv, raw_packet, primary_key) -> encryped_payload
aes_encrypt(iv, secondary_key) -> encrypted_iv
aes_encrypt(last_block, secondary_key) -> message_mac

*Decryption*:
aes_decrypt(encrypted_iv, secondary_key) -> iv
aes_cbc_decrypt(iv, encryped_payload, primary_key) -> raw_packet

Authentication (todo):
```
~~aes_decrypt(last_encrypted_block, secondary_key) -> transmitted_mac~~
~~aes_encrypt(last_encrypted_block, secondary_key) -> expected_mac~~
```
transmitted_mac from "header" in front of IV
aes_cbc_encrypt(decrypted_packet, key) to another buffer -> aes_encrypt (last_packet_of that, secondary key) -> expected_mac

if transmitted_mac != expected_mac
  drop_packet()
```

Please also note that this adds a 34 byte overhead to every packet (reducing this may be desiderable). A comparison of the results of the `benchmark` program available in n2n would also be useful to understand the performance impact of the changes.

emanuele-f commented on transform_aes.c:
```diff
@@ -22,15 +22,18 @@
+#define N2N_AES_TRANSFORM_VERSION       2  /* version of the transform encoding */
+#define N2N_AES_IVEC_SIZE               16 /* Enough space for biggest AES ivec */
+#define N2N_AES_MSGAUTH_SIZE		16
+#define HASH_SIZE			64
```
64 is because of SHA512, please add a comment

Logan007 replied:
Okay, I will add a comment. Alternatively, I could just use SHA512_DIGEST_LENGTH throughout the code without definition here - found out about it later.

Logan007 replied:
Used SHA512_DIGEST_LENGTH.

emanuele-f commented on transform_aes.c:
```diff
@@ -162,28 +167,23 @@ static ssize_t aes_choose_rx_sa( transop_aes_t * priv, const u_int8_t * peer_mac
- *  [V|SSSS|nnnnDDDDDDDDDDDDDDDDDDDDD]
- *         |<------ encrypted ------>|
+ *  [V|SSSS|MMMMMMMMMMMMMMMM|IIIIIIIIIIIIIIII|LLDDDDDDDDDDDDDDDDDDDDDDDDDDDDDD]
+ *         |< encrypted K2 >|< encrypted K2 >|<------- encrypted K1 --------->|
```
Replace "encrypted K2" with "encrypted with K2" or similar to avoid confusion

Logan007 replied:
I will think of something.

emanuele-f commented on transform_aes.c:
```diff
@@ -162,28 +167,23 @@ static ssize_t aes_choose_rx_sa( transop_aes_t * priv, const u_int8_t * peer_mac
  *  - a 32-bit SA number in clear text
- *  - ciphertext encrypted from a 32-bit nonce followed by the payload.
+ *  - a 128 bit inner message authentication code (last encrypted data block encrypted again using K2),
+ *  - the separately encrypted IV (using K2),
```
add size (128 bit)

Logan007 replied:
OK.

emanuele-f commented on transform_aes.c:
```diff
@@ -219,28 +220,37 @@ static int transop_encode_aes( n2n_trans_op_t * arg,
-            *pnonce = rand();
-            memcpy( assembly + TRANSOP_AES_NONCE_SIZE, inbuf, in_len );
+            data_len = (uint16_t *)assembly;
+            *data_len = in_len;
```
In order to make connection between big endian and little endian devices work properly, this must be encoded in network byte order. Check out https://en.wikipedia.org/wiki/Endianness for details. Please replace `*data_len = in_len;` with `*data_len = htons(in_len);`. By the way, why sending payload length required here?

Logan007 replied:
Thank you, I surely will change that.

The data length is required here in the first place because we need to keep the encrypted data "prefix free", meaning that no other ever encrypted message could be put in front of any other package encrypted data forming a valid packet again. To put packet's length upfront is an effective countermeasure to prevent this and therefore eliminate the possibility of a  variable-length attack described [here](https://en.wikipedia.org/wiki/CBC-MAC#Security_with_fixed_and_variable-length_messages).

emanuele-f commented on transform_aes.c:
```diff
@@ -219,28 +220,37 @@ static int transop_encode_aes( n2n_trans_op_t * arg,
+            len2 = ( ( ( len - 1) / AES_BLOCK_SIZE) + 1 ) * AES_BLOCK_SIZE; /* Round up to next whole AES block size */
+
+            RAND_bytes(&iv, 16);  /* get a cryptographically safely randomized initialization vector */
+            /* REVISIT: return value of RAND_bytes function should be 1 to continue cryptographically safe, otherwise exit */
```
Please add failure check. In this case, the communication should not proceed

Logan007 replied:
Sure.

emanuele-f commented on transform_aes.c:
```diff
@@ -219,28 +220,37 @@ static int transop_encode_aes( n2n_trans_op_t * arg,
+
+            RAND_bytes(&iv, 16);  /* get a cryptographically safely randomized initialization vector */
+            /* REVISIT: return value of RAND_bytes function should be 1 to continue cryptographically safe, otherwise exit */
+            memcpy((sa->enc_ivec), iv, 16 ); /* sa->enc_ivec gets changed during CBC encryption procedure, so keep a copy in iv */
```
Is it necessary to provide the IV on every packet? In my understanding, the IV is only required during the decryption of the very first block, is this correct?

Logan007 replied:
The IV is different for each packet. So, it must be provided once for every transmitted packet - at the beginning of the CBC decryption process.

It is generated and stored in "iv". Afterwards, it gets copied to enc_ivec. During the CBC decrpytion process, it seems that the content of enc_ivec gets changed with every single block inside the CBC crypto function. I guess this from some simple debugging showing that enc_ivec has a different content after the CBC function call. That's why we should keep it somewhere, here iv to use it later to put in the packet.

Ummm, maybe I could have directly generated it into the packet buffer and afterwards copy it into enc_ivec? How could byte order be correct?

Let me add that a random IV is not an issue. Au contraire! The encryption is necessary as the first data block is extremly predictable (preamble and MACs that are already known because transmitted in plain text as part of framing packet) - so we should not transmit IV in plain text and therefore scramble it. A possible attack as described [here](https://en.wikipedia.org/wiki/CBC-MAC#Allowing_the_initialization_vector_to_vary_in_value) is also encountered by the encryption of the IV.

Logan007 replied:
For Part 2 I already took a note to remove enc_iv an dec_iv from the sa structure and use local variables only. As IVs are varying for every packet, there is no need to keep track of them in the more static sa structure.

emanuele-f commented on transform_aes.c:
```diff
@@ -325,36 +338,34 @@ static int transop_decode_aes( n2n_trans_op_t * arg,
-                     * AES_BLOCKSIZE-1 */
-                    padding = assembly[ len-1 ] & 0xff; 
+                                     &(sa->dec_ivec), 0 /* decrypt */ );
+                    data_len = *(uint16_t*)assembly;
```
For the endianess, replace with: `data_len = ntohs(*(uint16_t*)assembly;)`

Logan007 replied:
Thank you!

emanuele-f commented on transform_aes.c:
```diff
@@ -325,36 +338,34 @@ static int transop_decode_aes( n2n_trans_op_t * arg,
+                    data_len = *(uint16_t*)assembly;
 
-                    if ( len >= (padding + TRANSOP_AES_NONCE_SIZE))
+                    // REVISIT: verifiy message authentication code by CBC'ing the decrypted block again and encrypting the last block using enc_key_2, and finally compare it to MSGAUTH in inbuf
```
Since the MAC has already been added to the packet, please add the remaining checks to validate it

Logan007 replied:
Okay, will try!

Logan007 replied:
I just realized that the message authentication does not comprise the version and security association number. I will fix that and create a new commit (3rd).

emanuele-f commented on transform_aes.c:
```diff
@@ -219,28 +220,37 @@ static int transop_encode_aes( n2n_trans_op_t * arg,
-            memset( &(sa->enc_ivec), 0, sizeof(N2N_AES_IVEC_SIZE) );
             AES_cbc_encrypt( assembly, /* source */
-                             outbuf + TRANSOP_AES_VER_SIZE + TRANSOP_AES_SA_SIZE, /* dest */
+                             outbuf + TRANSOP_AES_VER_SIZE + TRANSOP_AES_SA_SIZE + N2N_AES_MSGAUTH_SIZE + N2N_AES_IVEC_SIZE, /* dest */
```
Add a comment before each encryption/descryption function, explaining the operation performed (e.g. decrypting the iv, decrypting the packet payload)

Logan007 replied:
Sure.

emanuele-f commented on transform_aes.c:
```diff
@@ -219,28 +220,37 @@ static int transop_encode_aes( n2n_trans_op_t * arg,
+                         outbuf + TRANSOP_AES_VER_SIZE + TRANSOP_AES_SA_SIZE + N2N_AES_MSGAUTH_SIZE, /* to the buffer */
+                         &(sa->enc_key_2) ); /* using the secondary key */
+            AES_encrypt (outbuf + len2 - AES_BLOCK_SIZE, /* encrypt the last cipher block again */
+                         outbuf + TRANSOP_AES_VER_SIZE + TRANSOP_AES_SA_SIZE, /* to use it as message authentication code */
```
Please provide links to online resources/explain why performing the operation on the last block is enough for the authentication of the whole packet

Logan007 replied:
As CBC always uses the cipher of the block before to encrypt the next one, every cipher always depends on the block before and all other blocks before. Thus, the last block is a fingerprint of the whole encrypted data. Anyway, we can't just use it but need to encrypt it again using another key to prevent an attack described [here](https://en.wikipedia.org/wiki/CBC-MAC#Using_the_same_key_for_encryption_and_authentication).

Logan007 commented:
Thank you for the feedback! I give additional comments as you suggested in reply to your inline comments above.

Authentication verification should happen differently than outlined above, because CBC decryption happens in "parallel", i.e. the last block does not change in decryption process if the first cipher block has altered, because the decryption of the last block only requires the ciper of the (n-1)th block. So, the scheme would look like:
```
*Authentication* (todo):
transmitted_mac from "header" in front of IV
aes_cbc_encrypt(decrypted_packet, key) to another buffer -> aes_encrypt (last_packet_of that, secondary key) -> expected_mac

if transmitted_mac != expected_mac
  drop_packet()
```
By the way, a few bytes per packet could be saved here if the message authentication code only contained the maybe first 32 bit of the two-times-encrypted last bock. That is a choice to make with a view to the desired level of security.

Still learning git, I wonder how I can make the additional changes to the code as suggested - your comments are very valuable. New Pull Request? Please advice.

emanuele-f commented:
Thank you, I've updated the comment above for the authentication part. In order to update this pull request you only need to commit the new changes to your branch. So, after making the changes, do a `git add ...your_changed_files...`, `git commit` and then push the changes to your online branch with `git push`. This will automatically update the pull request, since your online branch is now *bound* to this pull request

Commit:
The issues from the review have been addressed.

Logan007 commented:
I have just finished implementing the changes you suggested and applied some obvious corrective changes.

You are right, concerning packet size, we now see larger packets: I would just put "minus 1" for dropping the padding length and "minus 4" for replacing the nonce. However, to keep it a little bit more flexible, I implemented the message authentication in a way that allows to change its length in bytes at compile time - just change the definition (0 ... 16, all seem to work well). As the length of the message authentication code is not explicitly transmitted, a change in message authentication code size causes incompatibility to clients with a different size. Maybe something to ponder for Part 2...

For now, there is just the need to agree on a reasonable value for it.

Commit:
Message Authentication now also includes aes version and sa number.

Logan007 commented:
The aes_version and sa number now are included in the message authentication by just xor'ing them before the message authentication code gets encrypted using K2. So, the accordingly updated message authentication verification scheme should look like this:

```
**transmitted_mac** from "header" in front of IV

aes_cbc_encrypt(decrypted_packet, key) to another buffer
-> xor'ing in the aes_version and sa field from the received packet to the last block
-> aes_encrypt (last_block_of that, secondary key) 
-> **expected_mac**

if transmitted_mac != expected_mac
  drop_packet()
```

Leftovers for Part 2 of AES revamp: Encounter Replay Attacks by modifying IV scheme (could remain compatible to this aes version) and clean up of sa structure.

I did some benchmarking (i7-5775C). The modifications indeed cost a bit of performance as follows:

Before modifcation (6.5 x slower than null):
```
./benchmark 
..........
[transop_null] 10000 times: (5255 -> 525 nsec each) 1551993236.392130 -> 1551993236.397385.
..........
[transop_twofish] 10000 times: (269364 -> 26936 nsec each) 1551993236.397415 -> 1551993236.666779.
..........
[transop_aes] 10000 times: (34192 -> 3419 nsec each) 1551993236.666785 -> 1551993236.700977.
```

After modification (15 x slower than null)
```./benchmark 
..........
[transop_null] 10000 times: (4033 -> 403 nsec each) 1551993168.241907 -> 1551993168.245940.
..........
[transop_twofish] 10000 times: (267498 -> 26749 nsec each) 1551993168.245970 -> 1551993168.513468.
..........
[transop_aes] 10000 times: (62269 -> 6226 nsec each) 1551993168.513472 -> 1551993168.575741.


```

So the gain of AES's hardware acceleration is melting a bit, but it is still way ahead of TF. Possible shortcuts could be taken by later maybe adding a command line parameters for fixed, lower keysize and disabled message authentication. Though, I would not recommend that at all from a security point of view.

Commit:
Added replay protection to IV.

Logan007 commented:
Sorry, if this pull request now gets a little bit lengthy... I have just added replay protection by encoding a  time stamp to the initialization vector as outlined in #73.

There are two values to parametrize the replay protection (definitions at the top of transform_aes.c):

- One is the time frame in which packets with a certain time stamp are accepted by the receiver by comparing the timestamp applied by the sender to the receiver's system time. This is set to +/- 16 seconds, allowing for 32 seconds of time difference between the two clocks plus transmission time. In the age of ubiquitious NTP servers, this might be reasonable.
- Nodes keep track of the previous packet's time stamp and make sure, that only packets with higher time stamp are accepted - assuming that packets do not overtake each other on the line too often. But as exactly this seems to happen from time to time, there is a grace time set to 160 ms which seems to work well in most cases, at least in the author's setup.

Also, I changed the message authentication code length to 4 bytes. Would that be a reasonable value?

The replay prevention does not come for free: There is a slight performance slump as follows (15.8 x slower than null, see the post above):

```
./benchmark
..........
[transop_null] 10000 times: (3732 -> 373 nsec each) 1552294709.652039 -> 1552294709.655771.
..........
[transop_twofish] 10000 times: (268661 -> 26866 nsec each) 1552294709.655796 -> 1552294709.924457.
..........
[transop_aes] 10000 times: (74913 -> 7491 nsec each) 1552294709.924461 -> 1552294709.999374.
```

By the way, talking performance, I found that key size has a somewhat non-neglectable impact on performance. Changing the now hard-wired keysize from 256 to 128 bit makes it look as follows (1400 ns faster):

```
./benchmark
..........
[transop_null] 10000 times: (3230 -> 323 nsec each) 1552293781.135773 -> 1552293781.139003.
..........
[transop_twofish] 10000 times: (270719 -> 27071 nsec each) 1552293781.139030 -> 1552293781.409749.
..........
[transop_aes] 10000 times: (60573 -> 6057 nsec each) 1552293781.409753 -> 1552293781.470326.
```

AES version incremented to 3. **This shall conclude the changes I wanted to apply for now.** Please let me know what you think and please test. Do the times for replay attack protection work for you? What does it take to make this pull request acceptable? I would love to hear back from you to make n2n even better.

A next step (new pull request, later) could include implementation of support for aes version 2 by iffing out the message authentication, the replay attack protection and also maybe switching to a lower keysize - allowing for reasonable use on less performing systems, e.g. routers.

Logan007 commented:
Author's evaluation after five days: works flawless in a setup with one public supernode and four edges around the world.

Logan007 commented:
I think there might be the need to make all the above more compatible with the existing dev branch versions. What do you think?

My thought is to add some automatism detecting the used version (1,2, or 3) per node by evaluating the first received packet and not allowing to go with less than that to prevent downgrade attacks. That way, the current dev version still would work with that. 

Also, I would like to hijack the "-A" option implemented by @emanuele-f . It could be extended to "-A 1", "-A 2", and "-A 3" to indicate the maximum AES packet version to use to send out from this edge node, still allowing other versions to be received. What would be a good default in case it is not given or a key file is used? 1 or 3? I might need a few hints to fully understand the mechanism how that information could get passed through to the transform_aes module.

Versions could be 

1. as in current dev (AES w/ CBC, 32 bit nonce, IV = 0) + [new] hashed password* + full key size* + length field in front of encrypted data* + randomized padding at data's end (not for packet length, if applicable)
2. as before + [new] full and random IV
3. as before + [new] replay attack protection

The three features marked with "*" could also be put to version 2 leaving version 1 nearly completely untouched and preserving full backward compatibility, if desired or necessary - which I would not recommend with a view to security as discussed in #68.

Worth further discussion?

emanuele-f commented:
Hi, this may be a good idea to specify different modes as this implementation requires more computation and leads to an increase in packet sizes. In my opinion the default mode should be a compromise between security and speed/size whereas a "paranoid" mode could be more security oriented. I would not add too much modes as it could confuse the users

Logan007 commented:
Okay, I guess we just need to think of them as protocol versions. Will try to implement the above sketched, merging version (2+3) to just (2). Might take a while...

Last night, I had an even more enhanced idea of how to further ameliorate security... will open a new issue with that.

Commit:
Keeping compatibility with AES encryption scheme version 1

Logan007 commented:
It took a while to restore backward compatibility towards the current dev branch aes encryption scheme which I shall call version 1.  All other changes to the AES scheme (random IV, message authentication, replay protection) now are bundled in encryption scheme version 2.

V2 is optional.  It can be activated for outgoing packets by using the "-A2" option instead of "-A" or "-A1" (new).  I was not able to get a blank between the A and the version number as it is an optional parameter argument for the sake of compatibility to "-A" only.  Incoming packets are always parsed, no matter what version.  That way, mixed networks are possible.

There were two things which I originally wanted to implement, but did not do so: 
1. The key length is still depending on the password size also in V2 to use full key spectrum.  That way, all 128 bit keys and 192 bit keys as well as 256 bit keys get a chance which enhances the usable key space.  By the way, for better entropy, V2 requires 4 password characters per one key byte.  So, to make use of a 256 bit key a password needs at least a 128 characters.  V1 remains unchanged in this respect.
2. A host based flag to keep track of the encryption scheme used was not implemented as there might be implications in key schedule mode which I could not foresee yet.  If it proves to be safe, it could be implemented anytime later without implications on the so far implemented schemes V1 or V2.

The code for encode and decode was completely linearized - I am sorry, Mr. [Dijkstra](https://homepages.cwi.nl/~storm/teaching/reader/Dijkstra68.pdf).  I did not see any reasonable way to indent error dependent parts of code across version related switch statements.  Some may find the code more readable now.

There is one drawback:  benchmark for AES is not working anymore - and I cannot really explain why. 
 Usually, I would not publish half-baked code, but in this case I hope for some hints or help.

Any thoughts?

Logan007 committed:
Merge branch 'dev' into revampAesPart1

Logan007 commented on edge_utils.c:
```diff
@@ -659,8 +655,11 @@ int edge_init_twofish_psk(n2n_edge_t * eee, uint8_t *encrypt_pwd,
 int edge_init_aes_psk(n2n_edge_t * eee, uint8_t *encrypt_pwd,
 		      uint32_t encrypt_pwd_len) {
+
+traceEvent (TRACE_ERROR, "!!! !!! !!! %d", eee->preferred_aes);
```
This will be removed during next commit. :)

emanuele-f commented on edge_utils.c:
```diff
@@ -36,7 +32,7 @@
 #endif /* #if defined(DEBUG) */
 
-#define REGISTER_SUPER_INTERVAL_MIN     5    /* sec */
+#define REGISTER_SUPER_INTERVAL_MIN     20   /* sec */
```
Please don't change in this pull request.

Logan007 replied:
I can't remember having changed that.  Maybe, that happened when I tried to have git resolve conflicts between latest dev and this pull request? I will put it manually to the state it should be.

emanuele-f commented on benchmark.c:
```diff
@@ -89,7 +89,7 @@ int main(int argc, char * argv[]) {
   memset(&transop_aes_cbc, 0, sizeof(transop_aes_cbc));
   transop_aes_init(&transop_aes_cbc);
-  transop_aes_setup_psk(&transop_aes_cbc, 0, encrypt_pwd, sizeof(encrypt_pwd)-1);
+  transop_aes_setup_psk(&transop_aes_cbc, 0, encrypt_pwd, sizeof(encrypt_pwd)-1,1);
```
It's better to use a define AES_SIMPLE_VERSION or AES_VERSION_1

Logan007 replied:
Good idea, I will add that - here and also throughout transform_aes.

emanuele-f commented on edge_utils.c:
```diff
@@ -1702,11 +1701,11 @@ const char *random_device_mac(void)
     }
 #ifdef WIN32
-#define random rand
+#define random() rand()
```
please explain the change

Logan007 replied:
Ummm, I honestly can't. Something weird happened to edge_utils. I will have a deeper look.

emanuele-f commented on edge_utils.c:
```diff
@@ -1755,10 +1754,8 @@ int quick_edge_init(char *device_name, char *community_name,
     setsockopt(eee.udp_multicast_sock, SOL_SOCKET, SO_REUSEPORT, &enable_reuse, sizeof(enable_reuse));
-#endif
-
+    
```
you have removed code here, SO_REUSEPORT, undo

Logan007 replied:
I will check, see above.

emanuele-f commented on transform_aes.c:
```diff
@@ -162,7 +189,7 @@ static ssize_t aes_choose_rx_sa( transop_aes_t * priv, const u_int8_t * peer_mac
 static size_t aes_best_keysize(size_t numBytes)
 {
-    if (numBytes >= AES256_KEY_BYTES )
+    if (numBytes >= AES256_KEY_BYTES)
```
Since hashing is now performed on the key, if you think that AES256 would fit both the computational use case of a desktop pc as well as a single core SBC board, then you can always return AES256_KEY_BYTES here

Logan007 replied:
That was my first thought, too. But to remain completely compatible to current builds (what _now_ is treated as V1), I suggest to keep it, otherwise compatibility of current dev version "-A" and this pull request's "-A1" will break.

As we saw in the benchmarks above, AES with 128 bit key is still 23% faster. So, my idea was, to leave it to the user to decide: by choosing the password length one determines the key length an thus computational load.

Based on explanation from #68, a 256 bit key is space is better than a 128 bit key space only. During the last days, when I found that the three zeroed keys (128, 192, 256 bits respectively) seem to deliver different encryptions, I had the idea that using the full available keyspace (all 128, 192 **and** 256 bitted keys) would be more than just restricting to the full key size of 256 bit.

That is why I left this routine also for V2. An attacker would not know what size of key we have used, he would need to brute force all 128, 192 and 256 keys instead of all 256 bit keys only - which of course already are a lot.

emanuele-f commented on transform_aes.c:
```diff
@@ -196,62 +236,174 @@ static int transop_encode_aes( n2n_trans_op_t * arg,
+	    assembly[ len2-1 ]=(len2-len);
             traceEvent( TRACE_DEBUG, "padding = %u", assembly[ len2-1 ] );
+	    /* set IV to 0 for v1 */
+	    memset (iv, 0, N2N_AES_IVEC_SIZE);
```
Version 1 should have IV properly initialized, not 0. This must be fixed in a lightweight way.

Logan007 replied:
I left it initialized to 0  for the sake of compatibility. If design decision is, that a new program version does not necessarily need to be compatible, the user has to replace the software on all hardware platforms used. This might be an issue, as some might use other forks that seem specifically designed for their needs or other platforms, e.g. Hin2n for unrooted android. But that is a decision to make.

Basically, an IV should not repeat under the use of the same key. If this is not possible, it should remain constant. That's why it is chosen to be 0. Another constant pattern would be as good as that, too.

The 32 bit nonce (I reintroduced for V1 for compatibility) serves "somehow as a kind of IV" because it prepends the first data bock that gets encrypted in cipher block _chaining_ mode - but it only is 32 bit wide. A full "random" IV as in V2 requires full block size of 128 bit. So the question here is, how to specify "lightweight"?

If it is decided that compatibility is not an issue, for V1, we could remove the nonce and switch to a random IV instead - it must be sent with the packet anyway; encrypted with a "not the data encrytion key (K2)" as first data block is highly predictable and we cannot use the data key. To stay lightweight, we could use half block size (64 bit) and use some hash of the secret password to fill up another 64 bit to full block size. That way, we only need to transmit the random part, IV would be quite random, and the constant part of IV would be password dependent and not hardwired in the code. That might be a compromise taking all concerns into account to at least some extent. That still does not allow for replay attack protection as in V2.

emanuele-f commented on transform_aes.c:
```diff
@@ -196,62 +236,174 @@ static int transop_encode_aes( n2n_trans_op_t * arg,
+	       to only copy the 52 (least) significant bits leaving the other 12 random as before
+	       - converting in host byte order to make sure bitwise masking is applied correctly;
+	       shifting might be independet from byte order, though. */
+	    micro_seconds = be64toh(*(uint64_t*)(outbuf + TRANSOP_AES_VER_SIZE + TRANSOP_AES_SA_SIZE + N2N_AES_MSGAUTH_SIZE));
```
Please also consider this issue with 32bit seconds: https://en.wikipedia.org/wiki/Year_2038_problem . Since we have extra bits it's a waste not to use them and instead make they random. Moreover, do we really need microseconds accuracy here? What if the clocks of the two machines are not synchronized?

Logan007 replied:
I actually have thought about it. What will happen is that the time counter turns over to zero. A reboot/restart after that will fix it.

IVs should not just not repeat under the same key, they also should be as unpredictable as possible. That is achieved by 2 means: Using as much randomness as space allows (thus only using the neccessary amount of the highly predictable time bits) and the following encryption step using K1 before feeding the IV to the cbc encryption.

The time stamp part serves two purposes in replay attack protection:

1. It makes sure that the packet was sent not too long ago. Taking into account that packets may take some time and also, that clocks might be not completely in sync, there is a grace time of +/- 16 seconds (= 32 seconds) for comparison of _clocks of different computers_, specified at the top of transform_aes. It is user's deliberately made decision to use V2 by setting the parameter. The user should be aware that higher security requirements such as replay attack protection also have some boundary conditions such as somewhat synchronous clocks in this case, I will add a section to the man page. In times of ubiquitous ntp servers, I think this is a feasible approach. Let me know what you think.

2. By comparing the packet time stamps to each other, i. e. comparing time stamps from the _same clock_, we make sure that packets are in sequence. The microseconds are a kind of counter. To be specific, we make sure that they are "somewhat" in sequence as we allow for some grace time of 160 ms because packets sometimes may overtake each other on the line or even the packet stack. This time is defined at the top of transform_aes, too. In my worldwide setup, this value yielded good results. The purpose is to prevent fast replay attacks inside the 32 seconds window mentioned above. That's what we need the high accuracy for. If tests show that the 160 _milli_seconds order is a sufficent reasonable grace time dimension, we could lower the accuracy of the time stamp further by the factor of 128...1024  microseconds by shifting it 7...10 bits further to the right and allowing more space for random bits at the same time.

emanuele-f commented:
Thank you for the work so far. I have added some comments to the code. I'm still not so much convinced by the changes as they add a lot of complexity (both coding and conceptual) although of course solve real problems. I'd have preferred a prior discussion on the design of all of this (maybe via a more direct communication channel) as this is now in a state "accept all and trust me that it works or throw it away". We need other users feedback on this, and also a feedback of my colleague @lucaderi .

Logan007 commented:
This is not meant to be be "take it or leave it".   I will remain open to apply any changes suggested and for discussion or explanations of course.  Though, I too was wondering why nobody else participated in the discussion so far. 

Thank you for your comments. Tomorrow, I will have a closer look and explain more, if required.

I just want to contribute to the project where I could be of use.  You can trust me :)

_PS: Please consider this more like a "just take what you need and leave the rest"-option._

emanuele-f commented:
Probably the crypto aspects are obscure for a lot of people. Anyway as I said, your contribution is welcome, we only have to tune clarity, simplicity and performance requirements with the security aspect.

Logan007 commented:
@emanuele-f, I think a design decision is required. There are several options on the table that need to be discussed in terms of compatability, security, performance and flexibility. At the end, these options could be handled in parallel as differenct encryption scheme versions. My guess is that a "one size fits it all" will probably not work out to cover user's needs in the long term.

Again, there is plenty of time and no need to rush these changes in the dev or any another branch as they should be considered security _enhancements_. Being an active n2n user, I will be around for while. And as this is a field where I feel that I am somewhat able to contribute, I am glad for the opportunity to give back to the open source community whose software I usually use.

If you prefer a more direct way of discussion, we could meet via teamspeak in the public lounge of www.awesomebros.net during the weekend. What time would be good for you?

Logan007 commented:
So, any input on compatibility and whether n2n should be kept compatible to current aes implementation?

emanuele-f commented:
The -A parameter of n2n has only been introduced recently, and there is no need to keep compatibility for now, in particular as no stable version of n2n has been released with the -A support

Logan007 commented:
And what about compatibility to aes from keyfiles? Those keys also use the value '1' in the version field?

emanuele-f commented:
I think keyfiles are only used by very few people, if any. The initialization fix I made should have already broken the interoperability between old and new version.

Logan007 commented:
I guess the dev branch has developed too far away in the meantime. Just take any idea or whatever you might need from the code. In whatever way you plan to implement any of these features (IV, replay protection, MAC), I would recommend to keep it flexible and extensible in terms of security, e.g. keysize, IV length, MAC tag length.

Shall we close this?
